crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1793,6 +1796,43 @@ impl Client {
         &self,
         room: impl Into<create_room::v3::Request<'_>>,
     ) -> Result<room::Joined> {
+        let (room, _) = self.get_or_create_room(room).await?;
+        return Ok(room);
+    }
+
+    /// Get or create a room using the `RoomBuilder` and send the request.
+    ///
+    /// Sends a request to `/_matrix/client/r0/createRoom`, returns a [`Result`]
+    /// containing an instance of [`Joined`][room::Joined] if successful.
+    #[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/src/docs/sync_running.md"))]
+    /// # Arguments
+    ///
+    /// * `room` - The easiest way to create this request is using the
+    /// `create_room::Request` itself.
+    ///
+    /// Returns tuple (room, created)
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use matrix_sdk::Client;
+    /// # use matrix_sdk::ruma::api::client::room::{
+    /// #     create_room::v3::Request as CreateRoomRequest,
+    /// #     Visibility,
+    /// # };
+    /// # use url::Url;
+    ///
+    /// # use futures::executor::block_on;
+    /// # block_on(async {
+    /// # let homeserver = Url::parse("http://example.com").unwrap();
+    /// let request = CreateRoomRequest::new();
+    /// let client = Client::new(homeserver).await.unwrap();
+    /// assert!(client.create_room(request).await.is_ok());
+    /// # });
+    /// ```
+    pub async fn get_or_create_room(
```

Comment:
As per comment before, I don't think it's possible to get or create a new room. The hang you experienced might be because:
1. A sync loop was not running
2. You called this inside an event handler

If you want, a `get_or_create_dm()` method does make sense, but that's not required to get this merged.